Paper bins show the paper inside them (atomised)

cacogen opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

I've been checking in on #57412 every day for awhile now. Based on the lack of activity it doesn't look like any maintainer wants to review it. In addition to adding overlays to paper bins, it adds a lot of other miscellanous improvements. 

So this afternoon I spent a couple of hours atomising out only the code that adds the overlays, and improving it. It uses far fewer overlays now (which has the added bonus of enabling taller stacks of paper). I'm hoping less code to review with less potential performance implications will make a maintainer more likely to review it. The other things I added in that PR can be added later, in a separate pull request. 

I've tested this thoroughly and fixed any issues I came across.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
![](https://camo.githubusercontent.com/aaf453ffc26aab74885c4ba694773872a479e9437c9963fa46fa94c656a6d96f/68747470733a2f2f692e696d6775722e636f6d2f435a774935434c2e706e67)

Paper bins display their contents.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: cacogen
qol: Paper bins show the paper inside them
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

cacogen committed:
Adds overlays to paper bins

Atomised and with better performance

SuperNovaa41 commented:
All it takes is a single ping in discord to get them to review it

SuperNovaa41 commented:
You should probably close the other PR then too

LemonInTheDark commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -103,14 +123,18 @@
 	else
 		return ..()
 
+/obj/item/paper_bin/proc/at_overlay_limit()
+	if(overlays.len >= MAX_ATOM_OVERLAYS)
+		return TRUE
+	return FALSE
+
```
You can just return the check wrapped in () if you'd like

cacogen replied:
Thanks

LemonInTheDark commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -120,50 +144,92 @@
+
+	if(LAZYLEN(papers))
+		var/paper_number = 1
+		for(var/obj/item/paper/current_paper in papers)
```
why not make this a `for(var/i in 1 to papers.len)` sort of deal?

cacogen replied:
Did it, thanks

LemonInTheDark commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -120,50 +144,92 @@
+	if(LAZYLEN(papers))
+		var/paper_number = 1
+		for(var/obj/item/paper/current_paper in papers)
+			if(paper_number == LAZYLEN(papers) || paper_number % 8 == 0)
+				var/mutable_appearance/paper_overlay = mutable_appearance(current_paper.icon, current_paper.icon_state)
+				paper_overlay.color = current_paper.color
+				paper_overlay.pixel_y = paper_number/8 - 2 //gives the illusion of stacking
+				. += paper_overlay
+				if(paper_number == LAZYLEN(papers)) //this is our top paper
+					. += current_paper.overlays //add overlays only for top paper
+					if(istype(src, /obj/item/paper_bin/bundlenatural))
+						bin_overlay.pixel_y = paper_overlay.pixel_y //keeps binding centred on stack
+					if(bin_pen)
+						pen_overlay.pixel_y = paper_overlay.pixel_y //keeps pen on top of stack
+			paper_number++
```
Could you use continue to make this less indented?

cacogen replied:
Did it, thanks

LemonInTheDark commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -120,50 +144,92 @@
 	qdel(src)
 
 /obj/item/paper_bin/bundlenatural/attackby(obj/item/W, mob/user)
+	if(istype(W, /obj/item/paper/carbon) && (W.icon_state == "paper_stack" || W.icon_state == "paper_stack_words"))
+		to_chat(user, "<span class='warning'>[W] won't fit into [src].</span>")
+		return
```
is there a way to achieve this without checking icon state names?

cacogen replied:
I'd prefer to compare the icon sizes but I'm not sure how to do that

LemonInTheDark replied:
yeah that's not possible with byond

cacogen replied:
I just refused all carbon paper, even though when the top sheet has been removed they're the correct width. Ideally I'd transform it to be one pixel less in width but I don't know how to do that

LemonInTheDark commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -120,50 +144,92 @@
-			CHECK_TICK
-		qdel(src)
+		if(W.use_tool(src, user, 1 SECONDS))
+			to_chat(user, "<span class='notice'>You slice the cable from [src].</span>")
```
Did you intend to add the cable stuff here? to be clear, I'm fine with it, I'd have merged your last pr, just want to make sure

cacogen replied:
Yeah, I did. I copied the whole proc over from the other PR to replace the while loop, which brought the use_tool over. It just makes it take a second.

maxymax13 commented:
reopen the last one 😕

cacogen committed:
Simplifies at_overlay_limit

Makes for() loop better

cacogen commented:
I'll make a separate PR with the stuff I left out of this one

cacogen committed:
Adds returns I forgot to avoid hitting overlay limit

cacogen committed:
Removes test bin

I spent awhile on the description but I think coders will think the object is self-evidently dumb

Mothblocks commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -15,19 +15,33 @@
-		update_appearance()
-	
+	if(mapload)
+		var/obj/item/pen/P = locate(/obj/item/pen) in src.loc
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Also, don't use `src.`. I know this is old code, but you should touch it up while you're here.

Mothblocks commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -15,19 +15,33 @@
+	update_appearance()
+
+/obj/item/paper_bin/proc/generate_paper()
+	var/obj/item/paper/P = new papertype(src)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -36,9 +50,23 @@
+		droppoint = drop_location()
+	if(collapse)
+		visible_message("<span class='warning'>The stack of paper collapses!</span>")
+	for(var/atom/movable/AM in contents)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -103,14 +125,16 @@
 		return ..()
 
+/obj/item/paper_bin/proc/at_overlay_limit()
+	return (overlays.len >= MAX_ATOM_OVERLAYS)
```
```suggestion
	return overlays.len >= MAX_ATOM_OVERLAYS
```

Mothblocks commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -120,50 +144,92 @@
+
+	if(LAZYLEN(papers))
+		for(var/paper_number in 1 to LAZYLEN(papers))
+			if(paper_number != LAZYLEN(papers) && paper_number % 8 != 0) //only top paper and every 8th paper get overlays
```
Move the 8 to a define, and replace both uses of it..

Mothblocks commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -120,50 +144,92 @@
+			var/obj/item/paper/current_paper = papers[paper_number]
+			var/mutable_appearance/paper_overlay = mutable_appearance(current_paper.icon, current_paper.icon_state)
+			paper_overlay.color = current_paper.color
+			paper_overlay.pixel_y = paper_number/8 - 2 //gives the illusion of stacking
```
Where's the 2 come from?

cacogen replied:
The paper overlays need to be shifted down 2 pixels so they are in the right position.

Mothblocks commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -120,50 +144,92 @@
+
+/obj/item/paper_bin/bundlenatural/Initialize(mapload)
+	binding_cable = new /obj/item/stack/cable_coil(src, 2)
+	binding_cable.color = "#a9734f"
```
Use the color defines.

Mothblocks commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -120,50 +144,92 @@
+
+/obj/item/paper_bin/bundlenatural/deconstruct(disassembled)
+	dump_contents()
+	qdel(src)
```
```suggestion
	return ..()
```

Mothblocks commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -60,38 +91,29 @@
-	else if(total_paper >= 1)
-		total_paper--
+	else if(LAZYLEN(papers))
+		var/obj/item/paper/top_paper = papers[LAZYLEN(papers)]
```
```suggestion
		var/obj/item/paper/top_paper = papers[papers.len]
```

It is guaranteed at this point that there are papers, using LAZYLEN here implies it could be 0.

Mothblocks commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -60,38 +91,29 @@
+		top_paper.add_fingerprint(user)
+		top_paper.forceMove(user.loc)
+		user.put_in_hands(top_paper)
+		to_chat(user, "<span class='notice'>You take [top_paper] out of \the [src].</span>")
```
```suggestion
		to_chat(user, "<span class='notice'>You take [top_paper] out of [src].</span>")
```

`\the` is implicit.

Mothblocks commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -36,9 +50,23 @@
-	if(total_paper)
-		total_paper = 0
+	if(LAZYLEN(papers))
+		papers.Cut()
```
```suggestion
		LAZYNULL(papers)
```

That way, the papers list goes back to null.

Mothblocks commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -36,9 +50,23 @@
+			AM.pixel_y = rand(-3,3)
+		if(!AM.pixel_x)
+			AM.pixel_x = rand(-3,3)
+	papers.Cut()
```
```suggestion
	LAZYNULL(papers)
```

cacogen committed:
Update code/modules/paperwork/paperbin.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

cacogen committed:
Update code/modules/paperwork/paperbin.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

cacogen committed:
Update code/modules/paperwork/paperbin.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

cacogen committed:
Update code/modules/paperwork/paperbin.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

cacogen committed:
Update code/modules/paperwork/paperbin.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

cacogen committed:
Update code/modules/paperwork/paperbin.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

cacogen committed:
does it

cacogen committed:
this

cacogen committed:
Sadly limits all carbon paper in natural paper bundles regardless of current sprite width

Mothblocks commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -15,30 +16,58 @@
-		for(var/i in papers)
-			qdel(i)
-		papers = null
+	if(LAZYLEN(papers))
+		for(var/paper in papers)
+			qdel(paper)
+		LAZYNULL(papers)
```
```suggestion
	QDEL_LIST(papers)
```

Works with lazy lists too.

cacogen replied:
Thanks, I like to learn things like this

Mothblocks commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -1,8 +1,9 @@
+#define PAPERS_PER_OVERLAY 8
```
Undef at the end.

Mothblocks commented on code/__DEFINES/colors.dm:
```diff
@@ -71,6 +71,7 @@
 
 #define COLOR_BROWN "#BA9F6D"
 #define COLOR_DARK_BROWN "#997C4F"
+#define COLOR_ANOTHER_BROWN "#a9734f"
```
Is it possible to describe this in relation to COLOR_BROWN and COLOR_DARK_BROWN? Is it lighter than COLOR_BROWN? If so, COLOR_LIGHTER_BROWN. For reference, look at the other names.

cacogen replied:
I actually did Google this for inspiration but couldn't find anything I thought fit. Closest I got was "strong orange(brown)/strong tangelo" which I just thought was confusing. I'll call it COLOR_ORANGE_BROWN.

Mothblocks commented:
Almost there!

cacogen committed:
Update code/modules/paperwork/paperbin.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

cacogen committed:
Undef at end

- ORANGE_BROWN
- Define for magic 2

cacogen committed:
Merge remote-tracking branch 'cacogen/paperbinoverlaysatomic' into paperbinoverlaysatomic

Mothblocks commented:
Undraft if ready.
